inc/Ui/Diff.php

```diff
@@ -2,502 +2,168 @@
-        $l_nav .= $form->toHTML();
-        //move forward
-        if ($l_next && ($l_next < $r_rev || !$r_rev)) {
-            $l_nav .= $this->diffViewlink('diffnextrev', $l_next, $r_rev);
         }
 
-        /*
-         * Right side:
-         */
-        $r_nav = '';
-        //move back
-        if ($l_rev < $r_prev) {
-            $r_nav .= $this->diffViewlink('diffprevrev', $l_rev, $r_prev);
-        }
-        //dropdown
-        $form = new Form(['action' => wl()]);
-        $form->setHiddenField('id', $ID);
-        $form->setHiddenField('rev2[0]', $l_rev);
-        $form->setHiddenField('difftype', $this->difftype);
-        $form->setHiddenField('do', 'diff');
-        $input = $form->addDropdown('rev2[1]', $r_revisions)->val($r_rev)->addClass('quickselect');
-        $input->useInput(false); // inhibit prefillInput() during toHTML() process
-        $form->addButton('do[diff]', 'Go')->attr('type','submit');
-        $r_nav .= $form->toHTML();
-        //move forward
-        if ($r_next) {
-            if ($pagelog->isCurrentRevision($r_next)) {
-                //last revision is diff with current page
-                $r_nav .= $this->diffViewlink('difflastrev', $l_rev);
-            } else {
-                $r_nav .= $this->diffViewlink('diffnextrev', $l_rev, $r_next);
-            }
+        // diff view type
+        if ($INPUT->has('difftype')) {
+            // retrieve requested $difftype
+            $this->preference['difftype'] = $INPUT->str('difftype');
         } else {
-            $r_nav .= $this->diffViewlink('diffbothnextrev', $l_next, $r_next);
+            // read preference from DokuWiki cookie. PageDiff only
+            $mode = get_doku_pref('difftype', $mode = null);
+            if (isset($mode)) $this->preference['difftype'] = $mode;
         }
-        return array($l_nav, $r_nav);
     }
 
     /**
-     * Create html link to a diff view defined by two revisions
+     * get extended revision info
      *
-     * @param string $linktype
-     * @param int $lrev oldest revision
-     * @param int $rrev newest revision or null for diff with current revision
-     * @return string html of link to a diff view
+     * @param int|string $rev  revision identifier, '' means current one
+     * @return array  revision info structure of a page or media file
      */
-    protected function diffViewlink($linktype, $lrev, $rrev = null)
+    protected function getExtendedRevisionInfo($rev)
     {
-        global $ID, $lang;
-        if ($rrev === null) {
-            $urlparam = array(
-                'do' => 'diff',
-                'rev' => $lrev,
-                'difftype' => $this->difftype,
+        $changelog =& $this->changelog;
+
+        if ($rev) {
+            $info = $changelog->getRevisionInfo($rev);
+        } elseif (file_exists($filename = $this->itemFN($this->id))) {
+            $rev = filemtime(fullpath($filename));
+            $info = $changelog->getRevisionInfo($rev) + array(
```

Comment:
I try to show a diff of a image. (http://wiebertje.dokuwiki.nl/start?do=media&ns= then select image, history and in history compare the first two items of the list (older do not generate the error))
In that case `$changelog->getRevisionInfo($rev)` returns a `false`.
PHP returns that the error 'Error: Unsupported operand types' because a false added up to an array.

```
2021-08-18 18:38:53	/home/gerrit/Sites/dokuwiki/dokuwiki/inc/Ui/Diff.php(144)	Error: Unsupported operand types
  #0 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/Ui/MediaDiff.php(210): dokuwiki\Ui\Diff->getExtendedRevisionInfo()
  #1 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/Ui/MediaDiff.php(116): dokuwiki\Ui\MediaDiff->showFileDiff()
  #2 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/media.php(1000): dokuwiki\Ui\MediaDiff->show()
  #3 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/template.php(1480): media_tab_history()
  #4 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/Ajax.php(235): tpl_mediaFileDetails()
  #5 /home/gerrit/Sites/dokuwiki/dokuwiki/inc/Ajax.php(24): dokuwiki\Ajax->callMediadetails()
  #6 /home/gerrit/Sites/dokuwiki/dokuwiki/lib/exe/ajax.php(22): dokuwiki\Ajax->__construct()
  #7 {main}
```
This works, but not sure if it is logic that $info is (sometimes) empty:
```php
            $info = $changelog->getRevisionInfo($rev);
            $info = (is_array($info) ? $info :[]) + array(
                'current' => true,
            );
```
If I make new edits, I cannot reproduce this case. 
Ah, founded: the expection is if there happens an external edit. Than the changelog has no entry available, and the `getRevisionInfo($rev)` returns false. So I think false is a valid return and e.g. the check `(is_array($info) ? $info :[]) is needed.